source/azure_iot_hub_client.c

```diff
@@ -0,0 +1,630 @@
+
+        /* Setup working buffer to be used by middleware */
+        pxAzureIoTHubClient->_internal.ulAzureIoTHubClientWorkingBufferLength =
+            ( azureiotUSERNAME_MAX + azureiotPASSWORD_MAX ) > azureiotTOPIC_MAX ? ( azureiotUSERNAME_MAX + azureiotPASSWORD_MAX ) : azureiotTOPIC_MAX;
+        pxAzureIoTHubClient->_internal.pucAzureIoTHubClientWorkingBuffer = pucBuffer;
+        pucNetworkBuffer = pucBuffer + pxAzureIoTHubClient->_internal.ulAzureIoTHubClientWorkingBufferLength;
+        ulNetworkBufferLength = ulBufferLength - pxAzureIoTHubClient->_internal.ulAzureIoTHubClientWorkingBufferLength;
+
+        /* Initialize Azure IoT Hub Client */
+        xHostnameSpan = az_span_create( ( uint8_t * ) pucHostname, ( int32_t ) ulHostnameLength );
+        xDeviceIDSpan = az_span_create( ( uint8_t * ) pucDeviceId, ( int32_t ) ulDeviceIdLength );
+
+        xHubOptions = az_iot_hub_client_options_default();
+
+        if( pxHubClientOptions )
+        {
+            xHubOptions.module_id = az_span_create( ( uint8_t * ) pxHubClientOptions->pucModuleID, ( int32_t ) pxHubClientOptions->ulModuleIDLength );
+            xHubOptions.model_id = az_span_create( ( uint8_t * ) pxHubClientOptions->pucModelID, ( int32_t ) pxHubClientOptions->ulModelIDLength );
+            xHubOptions.user_agent = az_span_create( ( uint8_t * ) pxHubClientOptions->pucUserAgent, ( int32_t ) pxHubClientOptions->ulUserAgentLength );
+        }
+        else
+        {
+            xHubOptions.user_agent = az_span_create( ( uint8_t * ) azureiothubUSER_AGENT, sizeof( azureiothubUSER_AGENT ) - 1 );
+        }
+
+        if( az_result_failed( xCoreResult = az_iot_hub_client_init( &pxAzureIoTHubClient->_internal.xAzureIoTHubClientCore,
+                                                                    xHostnameSpan, xDeviceIDSpan, &xHubOptions ) ) )
+        {
+            AZLogError( ( "Failed to initialize az_iot_hub_client_init: 0x%08x.", xCoreResult ) );
+            xResult = eAzureIoTHubClientInitFailed;
+        }
+        /* Initialize AzureIoTMQTT library. */
+        else if( ( xMQTTResult = AzureIoTMQTT_Init( &( pxAzureIoTHubClient->_internal.xMQTTContext ), pxTransportInterface,
+                                                    prvGetTimeMs, prvEventCallback,
+                                                    pucNetworkBuffer, ulNetworkBufferLength ) ) != eAzureIoTMQTTSuccess )
+        {
+            AZLogError( ( "Failed to initialize AzureIoTMQTT_Init: 0x%08x.", xMQTTResult ) );
+            xResult = eAzureIoTHubClientInitFailed;
+        }
+        else
+        {
+            pxAzureIoTHubClient->_internal.pucDeviceID = pucDeviceId;
+            pxAzureIoTHubClient->_internal.ulDeviceIDLength = ulDeviceIdLength;
+            pxAzureIoTHubClient->_internal.pucHostname = pucHostname;
+            pxAzureIoTHubClient->_internal.ulHostnameLength = ulHostnameLength;
+            pxAzureIoTHubClient->_internal.xAzureIoTHubClientTimeFunction = xGetTimeFunction;
```

Comment:
>xAzureIoTHubClientTimeFunction [](start = 43, length = 30)

nit: azureIoTHubclient is redundent in the name, possibly we can make it smaller like xGetTimeFunction

Reply:
we would probably have to change all the others like HMAC, token refresh, etc for parity sake

Reply:
Sure, but something like this:
https://github.com/Azure/azure-iot-middleware-freertos/blob/4ad026331a8b5c9642fb9f8bc91cf25b1af50a9c/source/include/azure_iot_provisioning_client.h#L75

Reply:
gotcha gotcha. done